### PR TITLE
Lcfs

### PR DESCRIPTION
**Pull request by lisehanebring**

### CODE REVIEW

File: include/DREAM/Equations/Fluid/LCFSLossRateTerm.hpp
Comment (vaoskar) [inline include/DREAM/Equations/Fluid/LCFSLossRateTerm.hpp:21]: len_t is unsigned, so if you want to set this variable to a negative value to indicate that there are no closed flux surfaces, you should change the type of this variable to int_t

File: src/Equations/Fluid/LCFSLossRateTerm.cpp
Comment (vaoskar) [inline src/Equations/Fluid/LCFSLossRateTerm.cpp:82]: A quite picky comment, but wouldn't it be possible to write lines 82-90 in a shorter way like

```
for(len_t ir = 0; ir < nr; ir++)
    if(this->sign * PsiDiff(ir)<0)
        this->ir_LCFS=(int_t)ir-1;
```
?

File: src/Equations/Fluid/LCFSLossRateTerm.cpp
Comment (vaoskar) [inline src/Equations/Fluid/LCFSLossRateTerm.cpp:109]: A very picky comment but you don't need the variable x here, you can just write the content of this function as

```
    if (ir > this->ir_LCFS) { 
        return = 1; 
    } else { 
        return = 0; 
    }
```

**Review by vaoskar — CHANGES_REQUESTED** [review]
Nice work! I only have a few very small and easy to fix-comments!

It did however occur to me that it might not have been intentional to put this implementation on the same branch as the hottail low-Z-implementation. Is the hottail low-Z-implementation ready to merge soon as well, or would it be worthwhile to remove it from this branch? I didn't review that part now as I'm not read up on how it works.

**Review by vaoskar — APPROVED** [review]
Looks good! You could just remove the outcommented lines and outdated comments I marked before merging. And then do whatever should be done with the hottail low-Z part and wait for Mathias approval.

File: src/Equations/Fluid/LCFSLossRateTerm.cpp
Comment (hoppe93) [inline src/Equations/Fluid/LCFSLossRateTerm.cpp:107]: Doesn't this routine return the difference between ``psi_p`` at ``ir+1/2`` and the edge ``psi_p``?

File: src/Equations/Fluid/LCFSLossRateTerm.cpp
Comment (hoppe93) [inline src/Equations/Fluid/LCFSLossRateTerm.cpp:114]: I would suggest renaming ``psi`` -> ``psi_f`` to clarify that it is actually an edge value of the poloidal flux.

File: src/Equations/Fluid/LCFSLossRateTerm.cpp
Comment (hoppe93) [inline src/Equations/Fluid/LCFSLossRateTerm.cpp:174]: Are you sure this is how you want to implement it in kinetic mode? This will just remove particles from a single point in $(p,\xi_0)$. Wouldn't it be more natural to make it uniform across all of momentum space? In that case, the term should have the same value everywhere, and that value just has to be scaled appropriately with $\mathcal{V}'\Delta p\Delta\xi_0$.

File: py/DREAM/Output/LCFSLoss.py
Comment (hoppe93) [inline py/DREAM/Output/LCFSLoss.py:27]: Shouldn't ``self.grid.r`` be replaced by the subset ``self.grid.r[r]``?

File: py/DREAM/Output/LCFSLoss.py
Comment (hoppe93) [inline py/DREAM/Output/LCFSLoss.py:52]: Are you sure you want to give ``r`` and ``t`` to ``calcScrapedOffJre()``  and not to ``plot()``. I think you want the latter, as otherwise the radial/time grid will be screwed up (same comment for ``plotScrapedOffIre`` below).

File: src/Equations/SPIHandler.cpp
Comment (hoppe93) [inline src/Equations/SPIHandler.cpp:330]: I suppose this is a borrow from the ``plasmoid_drift`` branch? But is it correct? Shouldn't ``ir`` be affected more directly? And do we really want these changes on this branch, before the ``plasmoid_drift`` branch is merged?

File: src/Settings/RunawaySourceTerms.cpp
Comment (hoppe93) [inline src/Settings/RunawaySourceTerms.cpp:169]: Are you sure this shouldn't be a ``bool`` instead?

File: py/DREAM/Settings/Equations/RunawayElectrons.py
Comment (hoppe93) [inline py/DREAM/Settings/Equations/RunawayElectrons.py:151]: Yeah, here it really looks like ``user_input_active`` should be a boolean instead!

**Review by hoppe93 — CHANGES_REQUESTED** [review]
Nice work! I have some questions and suggestions for improvements here and there. Overall, I think this can be merged very soon! :)

File: py/DREAM/Settings/Equations/IonSpecies.py
Comment (vaoskar) [reply py/DREAM/Settings/Equations/IonSpecies.py:75]: It is supposed to be used to set a non-zero ion transport already before the onset of the later exponentially decaying transport coefficients. It was passed to the co-argument in calcTransportCoefficientExpdecayAllChargedStates() (and then forwarded to calcTransportCoefficientExpdecaySingleChargedState()) before, but I would think that disappeared when you restructured the setting of the ion transport and merged it to this branch from master?

File: py/DREAM/Settings/Equations/IonSpecies.py
Comment (vaoskar) [reply py/DREAM/Settings/Equations/IonSpecies.py:1173]: see above

File: py/DREAM/Settings/Equations/Ions.py
Comment (vaoskar) [reply py/DREAM/Settings/Equations/Ions.py:66]: see above

File: src/Equations/SPIHandler.cpp
Comment (lvotta98) [reply src/Equations/SPIHandler.cpp:330]: Discuss!

File: py/DREAM/Settings/Equations/RunawayElectrons.py
Comment (hoppe93) [inline py/DREAM/Settings/Equations/RunawayElectrons.py:103]: These two lines are remnants from the hot-tail low-Z branch and should be deleted.

**Review by hoppe93 — CHANGES_REQUESTED** [review]
Nice work! I found two more lines from the hot-tail low-Z branch which should be deleted. We should also fix all of Oskar's comments before merging.

**Review by hoppe93 — APPROVED** [review]
Great! This looks ready to merge now!

**Pull request merged**
